### PR TITLE
Issue #67 The simplest possible condition syntax should be used

### LINKED ISSUE #67: Rule: The simplest possible condition syntax should be used
Replaces http://jira.sonarsource.com/browse/LINTVS-181

```
var x = a != null ? a : b;
if (a!=null) {x = a;} else {x = b;}
if (condition) {x = a;} else {x = b;}
x = (y==null)?new object():y;
```
Can be converted to
```
var x = a??b;
x = a??b;
x = condition?a:b;
x = y??new object();
```

requirement to use ternary only kicks in when both branches of a conditional assign value to same var. Only applies with simple if/else. A single `else if` would turn rule off

R# ID: *ConvertIfStatementToNullCoalescingExpression*, *ConvertIfStatementToConditionalTernaryExpression*, *ConvertConditionalTernaryToNullCoalescing*

Implements http://jira.sonarsource.com/browse/RSPEC-3240

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #67

### CODE REVIEW

**Comment by dbolkensteyn** [comment]
I got this error while editing the code:
```
        private static bool ExpressionCanBeNull(ExpressionSyntax expression, SemanticModel semanticModel)
        {
            var info = semanticModel.GetTypeInfo(expression).Type;
            return info.IsReferenceType ||
                   info.SpecialType == SpecialType.System_Nullable_T; // Here
        }
```

```
System.NullReferenceException was unhandled by user code
  HResult=-2147467261
  Message=Object reference not set to an instance of an object.
  Source=SonarLint
  StackTrace:
       at SonarLint.Rules.ConditionalSimplification.ExpressionCanBeNull(ExpressionSyntax expression, SemanticModel semanticModel) in C:\Users\dinesh\Desktop\sonarsource\sonarlint-vs\src\SonarLint\Rules\ConditionalSimplification.cs:line 340
       at SonarLint.Rules.ConditionalSimplification.<>c.<Initialize>b__12_0(SyntaxNodeAnalysisContext c) in C:\Users\dinesh\Desktop\sonarsource\sonarlint-vs\src\SonarLint\Rules\ConditionalSimplification.cs:line 82
       at SonarLint.Helpers.DiagnosticAnalyzerContextHelper.<>c__DisplayClass0_0`1.<RegisterSyntaxNodeActionInNonGenerated>b__0(SyntaxNodeAnalysisContext c) in C:\Users\dinesh\Desktop\sonarsource\sonarlint-vs\src\SonarLint\Helpers\DiagnosticAnalyzerContextHelper.cs:line 49
       at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__DisplayClass35_0`1.<ExecuteSyntaxNodeAction>b__1()
       at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock(DiagnosticAnalyzer analyzer, Action analyze)
  InnerException: 
```

File: src/SonarLint/Rules.Description/S3240.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<p>
+    In the interests of keeping code clean, the simplest possible conditional syntax should be used. That means using
+    the <code>??</code> operator for an assign-if-not-null operator, and using the ternary operator <code>?:</code> for
+    assignment to a single variable.
+</p>
+
+<h2>Noncompliant Code Sample</h2>
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules.Description/S3240.html]: @tvsonar I've slightly simplified & formatted the code examples

File: src/SonarLint/Rules/ConditionalSimplification.cs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    {
+        internal const string DiagnosticId = "S3240";
+        internal const string Title = "The simplest possible condition syntax should be used";
+        internal const string Description =
+            "In the interests of keeping code clean, the simplest possible conditional syntax should be used. That " +
+            "means using the \"??\" operator for an assign-if-not-null operator, and using the ternary operator \"?:\" " +
+            "for assignment to a single variable.";
+        internal const string MessageFormat = "Use the \"{0}\" operator here.";
+        internal const string Category = "SonarQube";
+        internal const Severity RuleSeverity = Severity.Major;
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/ConditionalSimplification.cs]: @tvsonar minor suits this rule better I believe - WDYT? (see http://confluence.internal.sonarsource.com/display/LANG/Coding+Rules+Guidelines#CodingRulesGuidelines-Defaultseverities)

File: src/SonarLint/Rules/ConditionalSimplification.cs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+                compared = binary.Left;
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool ExpressionCanBeNull(ExpressionSyntax expression, SemanticModel semanticModel)
+        {
+            var info = semanticModel.GetTypeInfo(expression).Type;
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/ConditionalSimplification.cs]: @tvsonar info can be `null` here

**Comment by dbolkensteyn** [comment]
@tvsonar does R# report on this case?

```
class MyClass
{
	public MyClass()
	{
		object x;
		x = null == RandomlyNull() ? b : RandomlyNull();
	}

	public object RandomlyNull()
	{
		return (new Random().Next() % 2 == 0) ? null : new object();
	}
}
```

**Comment by dbolkensteyn** [comment]
@tvsonar can the optimal codefix for the following snippet be suggested right-away, instead of doing it in steps?

```
if (a != null) // Noncompliant; needlessly verbose
{
  x = a;
}
else
{
  x = b;
}
```

File: src/SonarLint/Rules/ConditionalSimplification.cs (hunk 2 above)
Comment (tvsonar) [inline src/SonarLint/Rules/ConditionalSimplification.cs]: @dbolkensteyn Yep, minor seems better.

**Comment by tvsonar** [comment]
@dbolkensteyn No, R# doesn't report on this.

**Comment by dbolkensteyn** [comment]
LGTM - let's assume that relying on side effects in methods is a corner case and that the refactoring is safe.

**Pull request merged**